src/squlearn/util/qulacs/qulacs_circuit.py

```diff
@@ -0,0 +1,1029 @@
+                for i, operation in enumerate(relevant_operations):
+                    for j, param in enumerate(self.new_operators_used_parameters[iop][operation]):
+                        if param in gradient_parameters:
+                            outer_jacobian[i, gradient_param_dict[param]] = (
+                                self.new_operators_coeff_grad[iop][operation][j](*obs_param_list)
+                            )
+                outer_jacobians.append(outer_jacobian)
+            return outer_jacobians
+
+        # self._outer_jacobi_obs_cache[cache_value] = outer_jacobian
+
+        return outer_jacobian
+
+    def get_operators_for_gradient(
+        self,
+        gradient_parameters: Union[
+            None, ParameterVectorElement, List[ParameterVectorElement]
+        ] = None,
+    ):
+        """Returns the Qulacs observable function for the observable depending on parameters."""
+
+        if isinstance(gradient_parameters, ParameterVectorElement):
+            gradient_parameters = [gradient_parameters]
+        gradient_parameters = list(gradient_parameters) if gradient_parameters is not None else []
+
+        def observable_func(*args):
+
+            list_operators = []
+            for iop, observable in enumerate(self.new_operators):
+
+                relevant_operations = [
+                    i
+                    for i in range(len(observable))
+                    if any(
+                        param in gradient_parameters
+                        for param in self.new_operators_used_parameters[iop][i]
+                    )
+                ]
+
+                list_paulis = []
+                for op in relevant_operations:
+                    list_paulis.append(PauliOperator(observable[op], 1.0))
+                list_operators.append(list_paulis)
+
+            return list_operators
+
+        return observable_func
+
+
+def evaluate_circuit(circuit: QulacsCircuit, *args) -> np.ndarray:
+    """
+    Function to evaluate the Qulacs circuit with the given parameters.
+
+    Args:
+        circuit (QulacsCircuit): Qulacs circuit to evaluate
+        parameters (List[float]): List of parameters to evaluate the circuit
+
+    Returns:
+        np.ndarray: Result of the evaluation
+    """
+
+    # TODO: Check if this is desired functionality
+    # Collects the args values connected to the observable parameters
+    # obs_param_list = sum(
+    #     [
+    #         list(args[len(circuit._qualcs_gates_parameters) + i])
+    #         for i in range(len(circuit._qualcs_obs_parameters))
+    #     ],
+    #     [],
+    # )
+    if circuit._qualcs_obs_parameters:
+        obs_param_list = list(args[-1])
+    else:
+        obs_param_list = []
+
+    circ = circuit.get_circuit_func()(
+        *[
+            args[i]
+            for i, arg in enumerate(["param", "x"])
+            if arg in circuit._qualcs_gates_parameters
+        ]
+    )
```

Comment:
Hey @David-Kreplin, I made some changes here to allow for circuits without trainable parameters and features. I'm not entirely sure if this is still functioning as intended by you, so we should talk about it. I also think a solution using keyword args could be a lot cleaner and more robust. Same holds for all other evaluate functions later in the file